taichi/ir/ir.cpp

```diff
@@ -845,6 +839,28 @@ std::string AtomicOpExpression::serialize() {
   }
 }
 
+void AtomicOpExpression::flatten(FlattenContext *ctx) {
+  // replace atomic sub with negative atomic add
+  if (op_type == AtomicOpType::sub) {
+    val.set(Expr::make<UnaryOpExpression>(UnaryOpType::neg, val));
```

Comment:
Will this cause an circular reference?

Reply:
Good point, but nope. If you look into `set()`:

https://github.com/taichi-dev/taichi/blob/f5373b1395a66c78506de0fd3c172dffe0d444d0/taichi/ir/expr.h#L47-L49

It assigns `expr` to another `std::shared_ptr`. `shared_ptr::operator=()` will ref the new object, and deref the old object. https://en.cppreference.com/w/cpp/memory/shared_ptr/operator%3D